### PR TITLE
 Implement a mechanism for AMP integrations to request an identifier for the current client.

### PR DESCRIPTION
**Pull request by cramforce**
Implements large parts of #961.

Misses the following aspects from the spec:

- delegation of id generation to viewer.

For details see https://goo.gl/Mwaacs

### CODE REVIEW

File: src/cid.js [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+ * A base cid string value and the time it was last read / stored.
+ * @typedef {{time: number, cid: string}}
+ */
+let BaseCidInfo;
+
+
+class Cid {
+  /** @param {!Window} win */
+  constructor(win) {
+    /** @const  */
```
Comment (dvoytenko) [inline src/cid.js]: nit: space

File: src/cid.js [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+    this.win = win;
+
+    /** @private @const  */
+    this.sha384Base64_ = sha384Base64;
```
Comment (dvoytenko) [inline src/cid.js]: Why instance var?

File: src/cid.js [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+     * @private {?string}
+     */
+    this.baseCid_ = null;
+  }
+
+  /**
+   * @param {string} externalCidScope Name of the fallback cookie for the
+   *     case where this doc is not served by an AMP proxy.
+   * @param {!Promise} consent Promise for when the user has given consent
+   *     (if deemed necessary by the publisher) to usage of tracking
```
Comment (dvoytenko) [inline src/cid.js]: "for usage" or "to use"

File: src/cid.js [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+   *     identifier.
+   * @param {!Promise=} opt_persistenceConsent Dedicated promise for when
+   *     it is OK to persist a new tracking identifier. This could be
+   *     supplied ONLY by the code that supplies the actual consent
+   *     cookie.
+   *  @return {!Promise<?string>} A client identifier that should be used
+   *      within the current source origin and externalCidScope. Might be
+   *      null if no identifier was found or be made.
```
Comment (dvoytenko) [inline src/cid.js]: "could be made"?

File: src/cid.js [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+}
+
+/**
+ * Returns the "external cid". This is a cid for a specific purpose
+ * (Say Analytics provider X). It is unique per user, that purpose
+ * and the AMP origin site.
+ * @param {!Cid} cid
+ * @param {string} externalCidScope
+ * @param {!Promise} persistenceConsent
+ * @return {!Promise<string|null>}
```
Comment (dvoytenko) [inline src/cid.js]: ?string

File: src/cid.js [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+ */
+function getExternalCid(cid, externalCidScope, persistenceConsent) {
+  const url = parseUrl(cid.win.location.href);
+  if (!isProxyOrigin(url)) {
+    return new Promise(resolve => {
```
Comment (dvoytenko) [inline src/cid.js]: return Promise.resolve(...)?

File: src/cid.js [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+  return getBaseCid(cid, persistenceConsent).then(baseCid => {
+    return cid.sha384Base64_(
+        baseCid +
+        getSourceOrigin(url) +
+        externalCidScope);
+  });
+}
+
+/**
+ * Returns whether the URL has the origin of a proxy
```
Comment (dvoytenko) [inline src/cid.js]: period

File: src/cid.js [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+ *     factored into its own package.
+ */
+export function isProxyOrigin(url) {
+  return (url.origin == 'https://cdn.ampproject.org' ||
+      url.origin.indexOf('http://localhost:') == 0);
+}
+
+/**
+ * Returns the source origin of an AMP document for documents served
+ * on a proxy origin. Throws if the doc is not on a proxy origin.
```
Comment (dvoytenko) [inline src/cid.js]: "Throws an error" or "Fails"

File: src/cid.js [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+    return cid.sha384Base64_(
+        baseCid +
+        getSourceOrigin(url) +
+        externalCidScope);
+  });
+}
+
+/**
+ * Returns whether the URL has the origin of a proxy
+ * @param {Url} url URL of an AMP document.
```
Comment (dvoytenko) [inline src/cid.js]: !Url

File: src/cid.js [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+ */
+export function isProxyOrigin(url) {
+  return (url.origin == 'https://cdn.ampproject.org' ||
+      url.origin.indexOf('http://localhost:') == 0);
+}
+
+/**
+ * Returns the source origin of an AMP document for documents served
+ * on a proxy origin. Throws if the doc is not on a proxy origin.
+ * @param {Url} url URL of an AMP document.
```
Comment (dvoytenko) [inline src/cid.js]: !Url

File: src/cid.js [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+ * @return {string} The source origin of the URL.
+ * @visibleForTesting BUT if this is needed elsewhere it could be
+ *     factored into its own package.
+ */
+export function getSourceOrigin(url) {
+  // List of well known proxy hosts. New proxies must be added here
```
Comment (dvoytenko) [inline src/cid.js]: Sounds like this statement belongs in the `isProxyOrigin`?

File: src/cid.js [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+  const path = url.pathname.split('/');
+  const prefix = path[1];
+  assert(prefix == 'c' || prefix == 'v',
+      'Unknown path prefix in url %s', url.href);
+  const domainOrHttpsSignal = path[2];
+  const origin = domainOrHttpsSignal == 's'
+      ? 'https://' + path[3]
+      : 'http://' + domainOrHttpsSignal;
+  // Sanity test that what we found looks like a domain.
+  assert(origin.indexOf('.') > 0, 'Expected a . in origin %s Foo', origin);
```
Comment (dvoytenko) [inline src/cid.js]: Remove "Foo"?

File: src/cid.js [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+
+/**
+ * Returns the base cid for the current user. This string must not
+ * be exposed to users without hashing with the current source origin
+ * and the externalCidScope.
+ * On a proxy this value is the same for a user across all source
+ * origins.
+ * @param {!Cid} cid
+ * @param {!Promise} persistenceConsent
+ * @return {!Promise}
```
Comment (dvoytenko) [inline src/cid.js]: !Promise<?>

File: src/cid.js [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+ */
+function getBaseCid(cid, persistenceConsent) {
+  if (cid.baseCid_) {
```
Comment (dvoytenko) [inline src/cid.js]: So, we never care if it expires after it's read once. Right?

File: src/cid.js [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+  // Storing the value may require consent. We wait for the respective
+  // promise.
+  persistenceConsent.then(() => {
+    // The initial CID generation is inherently racy. First one that gets
+    // consent wins.
+    const relookup = read(win);
+    if (!relookup) {
+      store(win, newVal);
+    }
+  });
```
Comment (dvoytenko) [inline src/cid.js]: Do we need any warning reported if consent is denied?

File: src/cid.js [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+  return Promise.resolve(newVal);
```
Comment (dvoytenko) [inline src/cid.js]: Is this the right thing to return a value even if consent is denied? Or maybe we can reject the promise and use it as a signal that the recipient should not perform any action?

File: src/cid.js [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+  item['time'] = timer.now();
+  item['cid'] = cidString;
+  win.localStorage.setItem('amp-cid', JSON.stringify(item));
+}
+
+/**
+ * Retrieves a stored cid item from localStorage. Returns undefined if
+ * none was found
+ * @param {!Window} win
+ * @return {!BaseCidInfo}
```
Comment (dvoytenko) [inline src/cid.js]: {!BaseCidInfo|undefined}

File: src/cid.js [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+  const item = JSON.parse(val);
+  return {
+    time: item['time'],
+    cid: item['cid'],
+  };
+}
+
+/**
+ * Whether the retrieved cid object is expired and should be ignored.
+ * @param {{!BaseCidInfo}} storedCidInfo
```
Comment (dvoytenko) [inline src/cid.js]: too many curlies

File: src/cid.js [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+ */
```
Comment (dvoytenko) [inline src/cid.js]: \@return

File: src/cid.js [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+  const createdTime = storedCidInfo.time;
+  const now = timer.now();
+  return createdTime + BASE_CID_MAX_AGE_MILLIS < now;
+}
+
+/**
+ * Whether we should write a new timestamp to the stored cid value.
+ * We say yes if it is older than 1 day, so we only do this max once
+ * per day to avoid writing to localStorage all the time.
+ * @param {{!BaseCidInfo}} storedCidInfo
```
Comment (dvoytenko) [inline src/cid.js]: ditto on curlies and `@return`

File: src/cid.js [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+    // sha384 does not in practice, so we copy the values into
+    // a plain old array.
+    const array = new Array(16);
+    for (let i = 0; i < uint8array.length; i++) {
+      array[i] = uint8array[i];
+    }
+    return array;
+  }
+  // Support for legacy browsers.
+  return String(win.location.href + new Date().getTime() +
```
Comment (dvoytenko) [inline src/cid.js]: `timer.now()`

File: src/cid.js (hunk 1 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 2 above)
Comment (cramforce) [inline src/cid.js]: Added comment. For testing.

File: test/_init_tests.js [hunk 22]
```diff
@@ -60,6 +60,7 @@ afterEach(() => {
       console./*OK*/log(e);
     }
   }
+  window.localStorage.clear();;
```
Comment (dvoytenko) [inline test/_init_tests.js]: two `;`

File: src/cid.js (hunk 3 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 4 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 5 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 6 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 7 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 9 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 8 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 10 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 11 above)
Comment (cramforce) [inline src/cid.js]: Good catch!

File: src/cid.js (hunk 12 above)
Comment (cramforce) [inline src/cid.js]: Done :)

File: src/cid.js (hunk 13 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js [hunk 23]
```diff
@@ -0,0 +1,284 @@
...
+ * Stores a new cidString in localStorage. Adds the current time to the
+ * stored value.
+ * @param {!Window} win
+ * @param {string} cidString Actual cid string to store.
+ */
+function store(win, cidString) {
+  const item = {};
+  item['time'] = timer.now();
+  item['cid'] = cidString;
+  win.localStorage.setItem('amp-cid', JSON.stringify(item));
```
Comment (dvoytenko) [inline src/cid.js]: How does this get implemented? 

> it gets deleted when the user deletes their cookies.

File: src/cid.js (hunk 14 above)
Comment (cramforce) [inline src/cid.js]: Right. If you keep your browser tap open for a year I think that is fine to count it as a single session.

File: src/cid.js (hunk 15 above)
Comment (cramforce) [inline src/cid.js]: Not here I'd say.

Currently there is no deny, just non-consent (since the dialog is non-modal).

File: src/cid.js (hunk 23 above)
Comment (dvoytenko) [inline src/cid.js]: Got it.

File: src/cid.js (hunk 16 above)
Comment (cramforce) [inline src/cid.js]: The problem is that this value is needed in the call to check whether consent is needed/has been given already.

In practice (and I think I documented that in the API), the consent promise should actually depend on persistenceConsent-promise, so no other ids are issued until it resolved, but the one call to produce the persistenceConsent in the first place needs to call the API with a resolved promise for the consent.

Comment (cramforce) [inline src/cid.js]: Added some more docs to API.

File: src/cid.js (hunk 17 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 18 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 19 above)
Comment (cramforce) [inline src/cid.js]: Done

File: src/cid.js (hunk 20 above)
Comment (cramforce) [inline src/cid.js]: Done

File: test/_init_tests.js (hunk 22 above)
Comment (cramforce) [inline test/_init_tests.js]: Done

File: src/cid.js (hunk 21 above)
Comment (cramforce) [inline src/cid.js]: Done

**Comment by cramforce** [comment]
Comments addressed. Thanks for the awesome review!

File: src/cid.js (hunk 16 above)
Comment (dvoytenko) [inline src/cid.js]: Ok. Sounds good.

**Comment by dvoytenko** [comment]
LGTM

**Pull request merged**